### PR TITLE
Update the AMP_Twitter_Embed_Handler class to support all of the custom data attributes supported by the amp-twitter tag

### LINKED ISSUE #3050: amp-twitter sanitiser doesn't respect data-lang or data-conversation attributes
Afternoon, I've been looking into the amp sanitiser for a project. I've noticed the `AMP_Twitter_Embed_Handler` class doesn't support the full range of attributes supported by the component: https://amp.dev/documentation/components/amp-twitter/

Would you accept this as a PR?

### PR DESCRIPTION
**Pull request by benbowler**
Re: issue #3050 I did some research and the Jetpack `tweet` shortcode supports the same attributes that the [amp-twitter](https://amp.dev/documentation/components/amp-twitter/) tag does.

I've updated the `AMP_Twitter_Embed_Handler` class to track an array of `$supported_data_attributes` which is based on [Twitter's current embed documentation](https://developer.twitter.com/en/docs/twitter-for-websites/embedded-tweets/guides/embedded-tweet-parameter-reference). This prevents arbitrary access to data attributes on the elements while allowing an easy method for adding new ones if the API is updated in future.

Those attributes are then passed to `AMP_DOM_Utils::create_node`.

Any questions let me know. I've been testing with these tweets on a WP page:

```html
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr">Miss you, can’t wait to see back in the starting 11</p>
<p>— Kurt Kufuor (@Kurt_Kufour) <a href="https://twitter.com/Kurt_Kufour/status/1115992642632941568?ref_src=twsrc%5Etfw">April 10, 2019</a></p>
</blockquote>
<p><script charset="utf-8" type="text/javascript" src="https://platform.twitter.com/widgets.js" async=""></script></p>
<blockquote class="twitter-tweet" data-lang="en" data-theme="dark" data-conversation="none">
<p dir="ltr">WELCOME BACK! it was great seeing u even if on the bench &lt;3</p>
<p>— Catlady Dee (@PsycheDALIAc) <a href="https://twitter.com/PsycheDALIAc/status/1115993152115105792?ref_src=twsrc%5Etfw">April 10, 2019</a></p>
</blockquote>
<p><script charset="utf-8" type="text/javascript" src="https://platform.twitter.com/widgets.js" async=""></script></p>
<blockquote class="twitter-tweet" data-lang="en" data-conversation="none">
<p dir="ltr">Good for you Joe. What a great time to be back! YNWA</p>
<p>— Yang Bertweet (YB) (@AbdFauzi) <a href="https://twitter.com/AbdFauzi/status/1116003307162689537?ref_src=twsrc%5Etfw">April 10, 2019</a></p>
</blockquote>
<p><script charset="utf-8" type="text/javascript" src="https://platform.twitter.com/widgets.js" async=""></script></p>
```

Fixes #3050.

### CODE REVIEW

**Comment by benbowler** [comment]
@googlebot I signed it!

File: includes/embeds/class-amp-twitter-embed.php [hunk 1]
```diff
@@ -229,15 +245,27 @@ private function create_amp_twitter_and_replace_node( $dom, $node ) {
 			return;
 		}
 
+		$additional_attributes = [];
+
+		foreach ( $this->supported_data_attributes as $attr ) {
```
Comment (westonruter) [inline includes/embeds/class-amp-twitter-embed.php]: As opposed to limiting the attributes to just the explicitly supported ones, why not just copy _everything_? Any unrecognized `data-*` attributes will just be ignored by `amp-twitter`, but they may have some other purpose on the original `blockquote` element that had them.

File: includes/embeds/class-amp-twitter-embed.php [hunk 2]
```diff
@@ -299,13 +327,11 @@ private function sanitize_embed_script( $node ) {
...
 		// Handle case where script is immediately following.
-		$is_embed_script = (
-			$next_element_sibling
+		$is_embed_script = ( $next_element_sibling
 			&&
 			'script' === strtolower( $next_element_sibling->nodeName )
 			&&
-			false !== strpos( $next_element_sibling->getAttribute( 'src' ), $script_src )
-		);
+			false !== strpos( $next_element_sibling->getAttribute( 'src' ), $script_src ) );
```
Comment (westonruter) [inline includes/embeds/class-amp-twitter-embed.php]: Is this change necessary?

File: includes/embeds/class-amp-twitter-embed.php (hunk 1 above)
Comment (benbowler) [reply includes/embeds/class-amp-twitter-embed.php]: Yes I can re-work it on that basis.

File: includes/embeds/class-amp-twitter-embed.php (hunk 2 above)
Comment (benbowler) [reply includes/embeds/class-amp-twitter-embed.php]: This was my auto-formatter in VS Code.

Comment (benbowler) [reply includes/embeds/class-amp-twitter-embed.php]: A slight difference in the linter interpretation of the standards, both are valid it seems.

**Comment by benbowler** [comment]
I'll work on the test this week and update here.

File: includes/embeds/class-amp-twitter-embed.php [hunk 3]
```diff
@@ -229,15 +229,31 @@ private function create_amp_twitter_and_replace_node( $dom, $node ) {
...
 		}
 
+		$additional_attributes = [];
+
+		if ( $node->hasAttributes() ) {
+			foreach ( $node->attributes as $attr ) {
+				$name  = $attr->nodeName;
+				$value = $attr->nodeValue;
+
+				if ( substr( $name, 0, 5 ) === 'data-' ) {
```
Comment (westonruter) [inline includes/embeds/class-amp-twitter-embed.php]: Small improvement: no need for the `$name` or `$value` attributes. Just re-use `$attr->nodeName` here and `$attr->nodeValue` below.

File: includes/embeds/class-amp-twitter-embed.php [hunk 4]
```diff
@@ -229,15 +229,31 @@ private function create_amp_twitter_and_replace_node( $dom, $node ) {
...
-				'height'       => $this->DEFAULT_HEIGHT,
-				'layout'       => 'responsive',
-				'data-tweetid' => $tweet_id,
-			]
+			array_merge(
+				[
+					'width'        => $this->DEFAULT_WIDTH,
+					'height'       => $this->DEFAULT_HEIGHT,
+					'layout'       => 'responsive',
+					'data-tweetid' => $tweet_id,
```
Comment (westonruter) [inline includes/embeds/class-amp-twitter-embed.php]: Optionally these attributes could be put above in `$additional_attributes` (but renamed `$attributes`) and then the `foreach` loop above can just add the `data-` attributes to it. Then here there would be no need for `array_merge()`. It could just be:

```php
$new_node = AMP_DOM_Utils::create_node( $dom, $this->amp_tag, $attributes );
```

File: includes/embeds/class-amp-twitter-embed.php (hunk 3 above)
Comment (westonruter) [inline includes/embeds/class-amp-twitter-embed.php]: I wonder: should this not just go ahead and copy _all_ attributes? What if someone has a stylesheet that is targeting `.twitter-tweet`. Should not the `class` persist? What about other attributes?

**Comment by benbowler** [comment]
Checking the failures now.

File: includes/embeds/class-amp-twitter-embed.php (hunk 3 above)
Comment (benbowler) [reply includes/embeds/class-amp-twitter-embed.php]: Done

Comment (benbowler) [reply includes/embeds/class-amp-twitter-embed.php]: Good point, I now loop through all attributes.

File: includes/embeds/class-amp-twitter-embed.php (hunk 4 above)
Comment (benbowler) [reply includes/embeds/class-amp-twitter-embed.php]: Done

**Comment by benbowler** [comment]
Got there this afternoon @westonruter, let me know if you need any further changes.
